sdk/core/azure-core-http-jdk-httpclient/src/main/java/com/azure/core/http/jdk/httpclient/implementation/AzureJdkHttpRequest.java

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+package com.azure.core.http.jdk.httpclient.implementation;
+
+import com.azure.core.http.HttpMethod;
+import com.azure.core.implementation.util.HttpHeadersAccessHelper;
+import com.azure.core.util.Context;
+import com.azure.core.util.Contexts;
+import com.azure.core.util.ProgressReporter;
+import com.azure.core.util.logging.ClientLogger;
+import reactor.core.Exceptions;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.http.HttpClient;
+import java.net.http.HttpHeaders;
+import java.net.http.HttpRequest;
+import java.time.Duration;
+import java.util.Optional;
+import java.util.Set;
+import java.util.TreeMap;
+
+import static java.net.http.HttpRequest.BodyPublishers.noBody;
+
+/**
+ * Implementation of the JDK {@link HttpRequest}.
+ * <p>
+ * Using this instead of {@link HttpRequest#newBuilder()} allows us to optimize some cases now allowed by the builder.
+ * For example, setting headers requires each key-value for the same header to be set individually. This class allows
+ * us to set all headers at once. And given that the headers are backed by a {@link TreeMap} it reduces the number of
+ * String comparisons performed.
+ */
+public final class AzureJdkHttpRequest extends HttpRequest {
+    private final BodyPublisher bodyPublisher;
+    private final String method;
+    private final URI uri;
+    private final HttpHeaders headers;
+
+    /**
+     * Creates a new instance of the JDK HttpRequest.
+     *
+     * @param azureCoreRequest The Azure Core request to create the JDK HttpRequest from.
+     * @param context The context of the request.
+     * @param restrictedHeaders The set of restricted headers.
+     * @param logger The logger to log warnings to.
+     */
+    public AzureJdkHttpRequest(com.azure.core.http.HttpRequest azureCoreRequest, Context context,
+        Set<String> restrictedHeaders, ClientLogger logger) {
+        HttpMethod method = azureCoreRequest.getHttpMethod();
+        ProgressReporter progressReporter = Contexts.with(context).getHttpRequestProgressReporter();
+
+        this.method = method.toString();
+        this.bodyPublisher = (method == HttpMethod.GET || method == HttpMethod.HEAD)
+            ? noBody() : BodyPublisherUtils.toBodyPublisher(azureCoreRequest, progressReporter);
+
+        try {
+            uri = azureCoreRequest.getUrl().toURI();
+        } catch (URISyntaxException e) {
+            throw logger.logExceptionAsError(Exceptions.propagate(e));
+        }
+
+        this.headers = HttpHeaders.of(new HeaderFilteringMap(
+            HttpHeadersAccessHelper.getRawHeaderMap(azureCoreRequest.getHeaders()), restrictedHeaders, logger),
+            (ignored1, ignored2) -> true);
```

Comment:
is it possible to use this predicate to filter out restricted headers instead of `HeaderFilteringMap`?

Reply:
The BiPredicate is called for each value of the header, so if we had `restrictedHeader: val1,val2` it would log twice as there is two values.